File: project/project4.py
from PIL import Image
import numpy as np
import time
import skimage.color
from skimage.segmentation import mark_boundaries, find_boundaries
from skimage.segmentation import slic
import imageio
from SNIC import snic

# load image
image = np.array(Image.open('project4/white-tower.png'))
image1 = np.array(Image.open('project4/wt_slic.png'))
lab_image = skimage.color.rgb2lab(image)

# SNIC parameters
grid_of_x = 10
grid_of_y = 10
compactness = 40
iteration = 10

t1 = time.time()
[label_map_snic, centroids_snic] = snic(lab_image, grid_of_x, grid_of_y, compactness,)
t2 = time.time()
label_map_snic = label_map_snic.astype(int)
image_seg_snic = mark_boundaries(image, np.array(label_map_snic))
bd_snic = find_boundaries(np.array(label_map_snic)) *1
# img_uint8 = image_seg_snic.astype(np.uint8)
imageio.imwrite('tower_snic.png', image_seg_snic)
print('SNIC cost :', t2-t1)

#SLIC used skimage
segments = slic(image, n_segments = 100, compactness = 40)
t3 = time.time()
slic_skimage = mark_boundaries(image, segments)
bd_slic_skimage = find_boundaries(np.array(segments)) *1
# img1_uint8 = slic_skimage.astype(np.uint8)
t4 = time.time()
imageio.imwrite('tower_originalslic.png', slic_skimage)
print('Original SLIC cost:', t4-t3)

File: project/project4.m
clear all;

% SLIC
folder = 'project4/';
image=imread([folder 'white-tower.png']);
h2=size(image,1);
w2=size(image,2);
image=double(image);

% parameter
s=50;

% slic
image1=SLIC(image,s);

image1=uint8(image1(:,:,1:3));
figure,imshow(image1);
imwrite(image1,[folder 'tower_SLIC1.jpg']);

% color the pixel that touch two different clusters black
simage=image1;
for i=2:h2-1
    for j=2:w2-1
        flag=0;
        for t=1:3
            if image1(i,j,t)~=image1(i+1,j,t)
                flag=1;
            end
            if image1(i,j,t)~=image1(i,j+1,t)
                flag=1;
            end
            if image1(i,j,t)~=image1(i-1,j,t)
            	flag=1;
            end
            if image1(i,j,t)~=image1(i,j-1,t)
                flag=1;
            end
            if image1(i,j,t)~=image1(i+1,j+1,t)
                flag=1;
            end
            if image1(i,j,t)~=image1(i-1,j-1,t)
                flag=1;
            end
            if image1(i,j,t)~=image1(i+1,j-1,t)
                flag=1;
            end
            if image1(i,j,t)~=image1(i-1,j+1,t)
                flag=1;
            end
        end
        if flag==1
            simage(i,j,:)=[0 0 0];
        end
    end
end

figure,imshow(simage);
imwrite(simage,[folder 'tower_SLIC2.jpg']);